xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxClient.cpp

```diff
@@ -27,120 +27,38 @@
 
 #define FF_MAX_EXTRADATA_SIZE ((1 << 28) - FF_INPUT_BUFFER_PADDING_SIZE)
 
-
-class CDemuxStreamClientInternal
-{
-public:
-  CDemuxStreamClientInternal()
+CDemuxStreamClientInternal::CDemuxStreamClientInternal()
   : m_parser(nullptr)
   , m_context(nullptr)
   , m_parser_split(false)
   {
   }
 
-  ~CDemuxStreamClientInternal()
-  {
-    DisposeParser();
-  }
-
-  void DisposeParser()
-  {
-    if (m_parser)
-    {
-      av_parser_close(m_parser);
-      m_parser = nullptr;
-    }
-    if (m_context)
-    {
-      avcodec_close(m_context);
-      m_context = nullptr;
-    }
-  }
-
-  AVCodecParserContext *m_parser;
-  AVCodecContext *m_context;
-  bool m_parser_split;
-};
-
-class CDemuxStreamVideoClient
-: public CDemuxStreamVideo
-, public CDemuxStreamClientInternal
-{
-public:
-  CDemuxStreamVideoClient() {}
-  virtual std::string GetStreamInfo() override;
-};
-
-class CDemuxStreamAudioClient
-: public CDemuxStreamAudio
-, public CDemuxStreamClientInternal
-{
-public:
-  CDemuxStreamAudioClient() {}
-  virtual std::string GetStreamInfo() override;
-};
-
-class CDemuxStreamSubtitleClient
-: public CDemuxStreamSubtitle
-, public CDemuxStreamClientInternal
+CDemuxStreamClientInternal::~CDemuxStreamClientInternal()
 {
-public:
-  CDemuxStreamSubtitleClient() {}
-  virtual std::string GetStreamInfo() override;
-};
+  DisposeParser();
+}
 
-std::string CDemuxStreamVideoClient::GetStreamInfo()
+void CDemuxStreamClientInternal::DisposeParser()
 {
-  std::string strInfo;
-  switch (codec)
+  if (m_parser)
   {
-    case AV_CODEC_ID_MPEG2VIDEO:
-      strInfo = "mpeg2video";
-      break;
-    case AV_CODEC_ID_H264:
-      strInfo = "h264";
-      break;
-    case AV_CODEC_ID_HEVC:
-      strInfo = "hevc";
-      break;
-    default:
-      break;
+    av_parser_close(m_parser);
+    m_parser = nullptr;
   }
-
-  return strInfo;
-}
-
-std::string CDemuxStreamAudioClient::GetStreamInfo()
-{
-  std::string strInfo;
-  switch (codec)
+  if (m_context)
   {
-    case AV_CODEC_ID_AC3:
-      strInfo = "ac3";
-      break;
-    case AV_CODEC_ID_EAC3:
-      strInfo = "eac3";
-      break;
-    case AV_CODEC_ID_MP2:
-      strInfo = "mpeg2audio";
-      break;
-    case AV_CODEC_ID_AAC:
-      strInfo = "aac";
-      break;
-    case AV_CODEC_ID_DTS:
-      strInfo = "dts";
-      break;
-    default:
-      break;
+    avcodec_close(m_context);
+    m_context = nullptr;
   }
-
-  return strInfo;
 }
 
-std::string CDemuxStreamSubtitleClient::GetStreamInfo()
+template <class T>
+class CDemuxStreamClientInternalSub : public CDemuxStreamClientInternal, public T
```

Comment:
The name "sub" is confusing. At the first glance I would assume "subtitle" but then I wonder why this is used for video/audio